### ISSUE TITLE
Design/improve/unifty `sample()` interface

### ISSUE THREAD

**Issue by xukai92** [TuringLang/Turing.jl#746]
Questions and my current thoughts
- How to pass initial state
  - Support `sample(..., init::Array{Union{Missing, Array{Float64,1}},1}=[missing])`
  - We flat `init`, and assign those variables which are not missed to `vi[spl]`
  - Constructing a correct `init` is the user's responsibility
- How to pass adaptation configuration
  - Take use of `AdvancedHMC.Adaptation`
  - `sample(..., adapt::Union{Nothing,AbstractAdaptor}=nothing)`
- How to pass continuing chain
  - `sample(..., resume::Union{Nothing,AbstractChain}=nothing)`
  - How do we resume an adapted sampler?
    - With the new interface of `AdvancedHMC.Adaptation`, the adapted sampler is just the types but not the adapter or any other thing. So if we can make the types in Turing to be 1:1 to those types, we can resume the adapted sampler.

Related issue:
- Provide a way to set initial state of HMC #571
- Better adaptation configuration interface #491
- Resume interface #566

**Comment by trappmartin** [comment]
Does it make sense that we differentiate between initial value and resume sampling? Why not get rid of the resume interface and only provide a clean way to set an initial value?

**Comment by xukai92** [comment]
Good idea! Then the only question leaves here is how to resume adapted sampler.

**Comment by yebai** [comment]
I think we can use [`keyword`](https://docs.julialang.org/en/v1/manual/functions/index.html#Keyword-Arguments-1) arguments here to allow maximum flexibility and minimal `sample` design. For example:

```julia

function sample(rng::AbstractRNG, ℓ::Model, s::SamplerType, N::Integer; args...) where SamplerType
     for i=1:N
        step(rng::AbstractRNG, ℓ::Model, s::Sampler, N::Integer; args...)
    end
end

```

where `args` is a keyword argument, some example uses are

- `θ₀::AbstractVector = zeros(D)`
- `adaptor::Adaptor = StanNUTSAdaptor()`
- `n_chains::Int = 1`
- `resume::AbstractChain = chain`

**Comment by yebai** [comment]
Related: https://github.com/TuringLang/Turing.jl/issues/723

**Comment by yebai** [comment]
@cpfiffer Does it make sense moving this `sample` function to `MCMCChains`. This would allow packages such as `AdvancedHMC` can depend on `MCMCChains`, but not depend on `Turing`.

**Comment by cpfiffer** [comment]
I don't see why not. The function looks pretty small, and everything's parametric so there's not really any dependencies to speak of. The JuliaStan people may have some opinions though, so allow me to summon the JuliaStan kingpin.

@goedman, we're working on something of an overhaul of the Turing internals for our 0.7 release and wanted to know what you think about including a function in MCMCChains like the following:

```julia
function sample(rng::AbstractRange, 
    ℓ::Model, s::SamplerType, 
    N::Integer; args...
) where {Model, SamplerType, TransitionType}
    t = Array{TransitionType, 1}(undef, N)

    for i=1:N
        t[i] = step(rng::AbstractRange, ℓ::Model, s::SamplerType, N::Integer; args...)
    end
    
    return Chains(t . . .)
end
```

It's designed such that [AdvancedHMC](https://github.com/TuringLang/AdvancedHMC.jl) and Turing can work independently of one another. I'd view it as basically another MCMCChains constructor. Do you have any issues with including this in MCMCChains?

**Comment by goedman** [comment]
That's perfectly ok.

**Comment by goedman** [comment]
I've actually started to look at the AHMC sampler would love to spend some time with it next week and try it out! If the performance is reasonably close to DynamicHMC that would be a huge leap forward!

**Comment by cpfiffer** [comment]
@yebai Does this interface need some additional components? I've added all the bits to a draft of MCMCChains that includes the addition of `AbstractSampler`, `AbstractModel`, and `AbstractTransition` types which I could see Turing inheriting from. I can take this out, I'm not very convinced as to their necessity, but it does unify the "interface" component of all this.

I'm not sure on what `AbstractTransition` should look like, though, and I couldn't find an example of what this might be. The function above also doesn't have a mechanism for the type to actually make it into the function. Should `AbstractModel` have some parametric type `AbstractModel{TransitionType}`? [This comment](https://github.com/TuringLang/Turing.jl/issues/723#issue-423302952) seems to indicate that the `TransitionType` is at the sampler level.

Additionally, many of the samplers have some initial setup of some kind that isn't caught well by the "iterate through each step" methodology above. I added some "empty functions" that specific samplers can overload if they need to. It would be really cool if this could catch the general cases for all the samplers.

For example:

```julia
function sample(rng::AbstractRNG,
    ℓ::ModelType, s::SamplerType,
    N::Integer; kwargs...
) where {ModelType<:AbstractModel,
    SamplerType<:AbstractSampler,
    TransitionType<:AbstractTransition} #Not sure how to get TransitionType into the function
    t = Array{TransitionType, 1}(undef, N)

    # Perform any necessary setup.
    sample_init!(rng, ℓ, s, N; kwargs...)

    # Step through the sampler.
    for i=1:N
        t[i] = step(rng::AbstractRNG, ℓ::ModelType, s::SamplerType, N::Integer; kwargs...)
    end

    # Wrap up the sampler.
    sample_end!(rng, ℓ, s, N; kwargs...)

   # Placeholder while I figure out what t looks like.
    return Chains(t)
end
```

In summary, this is what I see the sampling interface looking like:

- Samplers inherit from `AbstractSampler` or some subtype of `AbstractSampler`, like `Hamiltonian<:AbstractSampler`
- Models inherit from the `AbstractModel` type. This won't be too hard on the Turing side, as 
    ```julia
    struct Model{pvars, dvars, F, TData, TDefaults}
    ```
  would just become
    ```julia
    struct Model{pvars, dvars, F, TData, TDefaults} <: AbstractModel
    ```
- Samplers that need setup or wrap-up code overload their own versions of `sample_init!(rng, ℓ, s, N; kwargs...)`  and `sample_end!(rng, ℓ, s, N; kwargs...)`.
- I have no idea what to do with `TransitionType` or `AbstractTransition`.

I also have wrapper functions for the above that allow for the sampling of multiple chains, and `psample` which will run multiple chains on parallel processes if any workers are available, but I can't really test them well without the whole interface working.

Comments, pointers, and corrections welcomed.

**Comment by yebai** [comment]
Here are some relevant `sample` implementations in the current code base:

HMC, SGLD and SGHMC:
- https://github.com/TuringLang/Turing.jl/blob/d2a987dda8e3e9e4d9056ad3b76f842f8feafa88/src/inference/hmc.jl#L214-L224

Gibbs 
- https://github.com/TuringLang/Turing.jl/blob/d2a987dda8e3e9e4d9056ad3b76f842f8feafa88/src/inference/gibbs.jl#L70-L76

PGibbs
- https://github.com/TuringLang/Turing.jl/blob/d2a987dda8e3e9e4d9056ad3b76f842f8feafa88/src/inference/pgibbs.jl#L82-L87

SMC
- https://github.com/TuringLang/Turing.jl/blob/d2a987dda8e3e9e4d9056ad3b76f842f8feafa88/src/inference/smc.jl#L71

Stan
- https://github.com/TuringLang/Turing.jl/blob/d2a987dda8e3e9e4d9056ad3b76f842f8feafa88/src/utilities/stan-interface.jl#L34-L41

**Comment by yebai** [comment]
> I have no idea what to do with TransitionType or AbstractTransition.


@cpfiffer This is meant to replace the `Sample` type and `Sampler.Info`, see https://github.com/TuringLang/Turing.jl/issues/723 and comments above (copied from #723)


> Models inherit from the AbstractModel type. 

I think we can inherit `Distributions.Sampleable` or `Distributions.Distribution`, e.g.

```julia
struct Model{pvars, dvars, F, TData, TDefaults} <: Distributions.Sampleable
```

**Comment by cpfiffer** [comment]
I've poked around a bit more and come up with an example of how I might think about modifying one of the samplers to fit this interface style. The changes below would fit fairly well into the common sampler interface function noted above, and give other potential Turing users a better example of how to construct modular sampler/model/general inference tools.

Using `SMC` as an initial example, I would imagine the following structure:

### The `Transition` Type

`ParticleContainer` seems like a natural thing to use as a `Transition` type. I'd think of this in terms of changing the [current definition](https://github.com/TuringLang/Turing.jl/blob/502027bd844ea61f081bef6a0f4dabc97fcd457c/src/core/container.jl#L77) of 

```julia
mutable struct ParticleContainer{T<:Particle, F}
  . . .
end
```

to simply inherit `AbstractTransition`, like so:

```julia
mutable struct ParticleContainer{T<:Particle, F} <: AbstractTransition
  . . .
end
```

The `SMC` package would then declare a function with the signature of 

```julia
MCMCChains.Chains(t::ParticleContainer)
```

which would handle the retrieval of all the samples. This would remove the need to generate a `Sample` type --- we'd just overload a `Chains` function for each `Transition` type to retrieve the samples using multiple dispatch. Not too much in the way of moving parts here.

An alternative specification for `Transition` would be to use the existing `step` function (which I'm not certain is currently used?) and save `vi.vals` each step. Then, the vector could be packaged up and sent off to MCMCChains.

https://github.com/TuringLang/Turing.jl/blob/d2a987dda8e3e9e4d9056ad3b76f842f8feafa88/src/inference/smc.jl#L48

It looks only slightly more complex for the `HMC` samplers. In this case we'd need to keep all the values from `Sample(vi, spl).vals` separately and then pull out varnames at the end. This would be doable during the `sample_end!` phase, or by making a function with the signature `MCMCChains.Chains(t::TransitionType, spl::HamiltonianSampler)`. The `spl` here is necessary to get the `vi` info to the `Chains` object.

### The `Model` type

As noted above, this should have to change too much:

```julia
struct Model{pvars, dvars, F, TData, TDefaults} <: Distributions.Sampleable
```

### The `Sampler` type

In order to remove `Sampler.info`, we could just declare special `Sampler` types for each sampler, and store each sampler's necessary fields in the struct itself. The current definition of the `SMC` sampler dumps some random stuff into the generic `Sampler` struct:

```julia
function Sampler(alg::SMC, s::Selector)
    info = Dict{Symbol, Any}()
    info[:logevidence] = []
    return Sampler(alg, info, s)
end
```

For reference, the `Sampler` struct is

```julia
mutable struct Sampler{T} <: AbstractSampler
    alg      ::  T
    info     ::  Dict{Symbol, Any} # sampler infomation
    selector ::  Selector
end
```

The way I might think about changing this set up is to remove the generic `Sampler` type (or maybe replace it with the most bare-bones version possible for interface purposes) and force each sampler type to create it's own type:

```julia
mutable struct SMCSampler{T} <: AbstractSampler
    alg :: T
    logevidence :: Float64
    selector ::  Selector
    particles ::  ParticleContainer{Trace} # 
end
```

The gist here would be to just move everything currently stored in any sampler's `spl.info` entry and put it directly into the struct. An overload of the `sample_init!` and `sample_end!` functions would be necessary, as well:

```julia
function sample_init!(::AbstractRNG,
        model::ModelType,
        s::SMCSampler,
        ::Integer;
        kwargs...
    ) where ModelType<:Distributions.Sampleable
    # Instantiate the ParticleContainer.
    s.particles = ParticleContainer{Trace}(model)
    push!(s.particles, spl.alg.n_particles, spl, VarInfo())
end
```

**Comment by yebai** [comment]
Thanks, @cpfiffer. The plan looks sensible. To summarise, the new design removes the generic `Sampler` type and requires each algorithm to (optionally) implement two types

- `SomeSampler <: AbstractSampler`
- `SomeTransition <: AbstractTransition`

It is worth also noting that only some algorithms make use of `Sampler.info`, for example

- `SMC`, `PG`, etc: `info[:logevidence]`
- `HMC`, `SGHMC`, etc: `info[:eval_num]` and `info[:i]`
- `Gibbs`, `IPMCMC`: info[:samplers]
- `MH`: `info[:proposal_ratio]`, `info[:prior_prob]`, `info[:violating_support]`

So perhaps we can provide a default `Sampler` implementation for algorithms that do not utilities `Sampler.info`. Then, we can explicitly require algorithms utilising `Sampler.info` to define a customised `Sampler` type, e.g.

```julia

struct Sampler <: AbstractSampler # Default for all algorithms

mutable struct Sampler{SMC} <: AbstractSampler
   log_evidence::Float64
end

```

Below are all `Sampler` definition in Turing. 

https://github.com/TuringLang/Turing.jl/blob/a4f78b61d3129329e197d15629104fd6e1f4c412/src/inference/AdvancedSMC.jl#L35-L39

https://github.com/TuringLang/Turing.jl/blob/a4f78b61d3129329e197d15629104fd6e1f4c412/src/inference/AdvancedSMC.jl#L114-L118

https://github.com/TuringLang/Turing.jl/blob/a4f78b61d3129329e197d15629104fd6e1f4c412/src/inference/gibbs.jl#L40-L68

https://github.com/TuringLang/Turing.jl/blob/a4f78b61d3129329e197d15629104fd6e1f4c412/src/inference/hmc.jl#L204-L211

https://github.com/TuringLang/Turing.jl/blob/a4f78b61d3129329e197d15629104fd6e1f4c412/src/inference/is.jl#L38-L41

https://github.com/TuringLang/Turing.jl/blob/a4f78b61d3129329e197d15629104fd6e1f4c412/src/inference/mh.jl#L51-L68

https://github.com/TuringLang/Turing.jl/blob/a4f78b61d3129329e197d15629104fd6e1f4c412/src/inference/AdvancedSMC.jl#L310-L340

https://github.com/TuringLang/Turing.jl/blob/a4f78b61d3129329e197d15629104fd6e1f4c412/src/inference/AdvancedSMC.jl#L507-L526

**Comment by cpfiffer** [comment]
One thing I'm a little fuzzy on is how the proposed sampler API works with compositional Gibbs, particularly as it relates to how you declare sample sizes in the components. Currently, you can do this:

```julia
chn = sample(model, Gibbs(1000, HMC(1, 0.2, 3, :x), PG(20, 1, :y)))
```

The sampler API forces the number of steps to be outside the `InferenceAlgorithm` contructor, and so we'd maybe have something that looks like

```julia
chn = sample(model, Gibbs(HMC(0.2, 3, :x), PG(20, :y)), 1000)
```

Is this an issue? I presume this would force `HMC` and `PG` to run one iteration apiece, but I don't know if the flexibility lost there is a problem or not.

**Comment by yebai** [comment]
It would be nice to have some support for running `HMC` or `PG` multiple steps in a Gibbs step. If the current API is troublesome (although I don't quite catch the issue @cpfiffer mentioned), we can re-design it. Following @cpfiffer's design, this would be adapted as:

```
chn = sample(model, Gibbs(HMC(0.2, 3, :x), PG(20, :y)), 1000) # default

chn = sample(model, Gibbs(HMC(0.2, 3, :x), PG(20, :y)), (1000,2,2)) # HMC - 2 steps, PG - 2 steps

```

**Comment by xukai92** [comment]
Or
```julia
# default
chn = sample(model, Gibbs(HMC(0.2, 3, :x), PG(20, :y), (1,1)), 1000) 

chn = sample(model, Gibbs(HMC(0.2, 3, :x), PG(20, :y), (2, 2)), 1000) 
```
?

**Comment by cpfiffer** [comment]
I like @xukai92's idea better, then there's no need to handle this weird edge case on the interface side, since `N` is strongly typed as an integer.

**Comment by yebai** [comment]
Fixed by #793
